Add support for XDG_CONFIG_HOME and AppData on Windows

samcoe opened the pull request:
This PR adds support for storing `gh` config files in `XDG_CONFIG_HOME` as well as `AppData` on Windows.

cc https://github.com/cli/cli/issues/2470
closes https://github.com/cli/cli/issues/1944
closes https://github.com/cli/cli/issues/554
supersedes https://github.com/cli/cli/pull/2080

samcoe committed:
Add support for XDG_CONFIG_HOME and AppData on Windows

samcoe commented on internal/config/config_file.go:
```diff
@@ -8,23 +8,71 @@ import (
+	} else if b := os.Getenv(XDG_CONFIG_HOME); b != "" {
+		path = filepath.Join(b, "gh")
+	} else if c, err := os.UserConfigDir(); runtime.GOOS == "windows" && err == nil && c != "" {
+		path = filepath.Join(c, "GitHub CLI")
```
Is this the folder name we want on Windows?

vilmibm replied:
I'm not an expert but it "feels" right given what I've seen among Windows applications.

mislav commented on internal/config/config_file.go:
```diff
@@ -8,23 +8,71 @@ import (
+	var path string
+	if a := os.Getenv(GH_CONFIG_DIR); a != "" {
+		path = a
+	} else if b := os.Getenv(XDG_CONFIG_HOME); b != "" {
```
Should we even respect XDG_CONFIG_HOME on Windows? I do not think we should, since XDG was not meant for Windows.

samcoe replied:
I was unsure of this. There were these comments (https://github.com/cli/cli/issues/2470#issuecomment-739575036, and https://github.com/cli/cli/issues/1944#issuecomment-716046710) from our users that pushed me to include support for it even on Windows.

mislav replied:
Ah, good point. I guess there is no strong reason to *not* support it. So if someone wants to pretend that they're on a Unix in their Windows, we can let them. 👍

mislav commented on internal/config/config_file.go:
```diff
@@ -8,23 +8,71 @@ import (
+		path = a
+	} else if b := os.Getenv(XDG_CONFIG_HOME); b != "" {
+		path = filepath.Join(b, "gh")
+	} else if c, err := os.UserConfigDir(); runtime.GOOS == "windows" && err == nil && c != "" {
```
TIL `os.UserConfigDir`! It sounds like a really useful function in general, but since we will only use its value on Windows, then I don't think it makes much sense to use it, because we're not utilizing most of its functionality. Let's just use `Getenv("AppData")` instead and cut out the middleman

mislav commented on internal/config/config_file.go:
```diff
@@ -8,23 +8,71 @@ import (
+func autoMigrateConfigDir(newPath string) {
+	var paths []string
+	p1, _ := os.UserHomeDir()
+	paths = append(paths, p1)
```
Nit: Could you only append this path if the error value when obtaining `p1` was nil?

mislav commented on internal/config/config_file.go:
```diff
@@ -8,23 +8,71 @@ import (
+	p1, _ := os.UserHomeDir()
+	paths = append(paths, p1)
+	p2, _ := homedir.Dir()
+	paths = append(paths, p2)
```
Ditto error handling

mislav commented on internal/config/config_file.go:
```diff
@@ -8,23 +8,71 @@ import (
+	}
+}
+
+var migrateConfigDir = func(oldPath, newPath string) {
```
Could there be a way to keep this part of the code "live" in tests, i.e. to avoid stubbing it?

Recent example: we used to stub `WriteConfigFile` (a pretty critical piece of functionality) in our tests until recently, but then we had a regression in writing the config file on Windows and our tests didn't catch it because we didn't exercise the filesystem operation. https://github.com/cli/cli/pull/3666

Could we instead have tests that set up/migrate actual config directories on disk, just switch them away from the user's actual home dir location?

mislav commented on internal/config/config_file.go:
```diff
@@ -8,23 +8,71 @@ import (
+	var paths []string
+	p1, _ := os.UserHomeDir()
+	paths = append(paths, p1)
+	p2, _ := homedir.Dir()
```
We moved away from `homedir.Dir()` because it can be a pretty expensive operation (take a look at its implementation; it does a lot!). Even though we keep backwards compatibility with it, we used to avoid invoking it unless it was the absolutely last filesystem location that we tried. The old `homeDirAutoMigrate` function was specifically designed around this approach. This change removes the guard and unconditionally calls `homedir.Dir()` even though other path in `p1` might be valid and might get used, ignoring the result of `p2`.

Couldn't we just keep using `homeDirAutoMigrate` intact and just migrate the result of that to the new location? I.e. to wrap the old migrate in a new migrate function.

samcoe replied:
That is good context. I was wondering why the implementation of `homeDirAutoMigrate` was designed that way. I can restore that behavior.

mislav reviewed (changes requested):
This is heading in a great direction. Thanks for working on this!

Left some comments about tightening up the migration logic and ensuring that it's entirely exercised in tests.

Note that if we advertise support for XDG spec, then we have to adhere to it as much as we can. That means avoiding storing non-config files under XDG_CONFIG_HOME. Currently, our only non-config file is `state.yml`, so that one should be migrated to `XDG_STATE_HOME` accordingly. https://github.com/cli/cli/issues/554#issuecomment-595803387

samcoe commented:
@mislav thanks for the thorough review. I addressed your comments. Regarding, `XDG_STATE_HOME` I was planning on addressing that next, in a separate PR.

angryTOASTpac reviewed (commented):
I'm fired up about this!

mislav commented on internal/config/config_file.go:
```diff
@@ -8,23 +8,76 @@ import (
+	}
+
+	// If the path does not exist try migrating config from default paths
+	if _, err := os.Stat(path); err != nil {
```
I think we should only try to migrate if `errors.Is(err, os.ErrNotExist)`

mislav commented on internal/config/config_file.go:
```diff
@@ -8,23 +8,76 @@ import (
-	return homeDir
+func dirExists(path string) bool {
+	f, err := os.Stat(path)
+	if err == nil && f.IsDir() {
+		return true
+	}
+	return false
```
```suggestion
	return err == nil && f.IsDir()
```

samcoe replied:
Nice catch

mislav commented on internal/config/config_file.go:
```diff
@@ -8,23 +8,76 @@ import (
+	return false
+}
+
+var migrateConfigDir = func(oldPath, newPath string) {
```
I wish that this wouldn't be an overridable function anymore, but I understand that other pieces of our codebase are coupled to ConfigDir in tests and that they could inadvertently cause changes to the user's home directory.

samcoe replied:
Yeah, unfortunately that is the case now. Hopefully in the future we can clean that up. I did add tests for this directly though so it isn't stubbed out in all tests.

mislav commented on internal/config/config_file_test.go:
```diff
@@ -150,20 +153,149 @@ func Test_parseConfigFile(t *testing.T) {
+		{
+			name: "GH_CONFIG_DIR specified",
+			env: map[string]string{
+				"GH_CONFIG_DIR": "/tmp/gh_config_dir",
```
On Windows, it's unlikely that a path set by environment variable would have slashes. It's more likely that it would have backslashes. You can normalize this via
```suggestion
				"GH_CONFIG_DIR": filepath.FromSlash("/tmp/gh_config_dir"),
```

mislav commented on internal/config/config_file.go:
```diff
@@ -8,23 +8,76 @@ import (
+	}
+
+	_ = os.MkdirAll(filepath.Dir(newPath), 0755)
+	_ = os.Rename(oldPath, newPath)
```
Should we also remove the orphaned `~/.config` directory after the migration of `~/.config/gh` to the new location?

We could simply try to `os.Remove` the parent. If it errors out, we ignore it.

samcoe replied:
I don't think this is a good idea since it could remove files users want to keep. I actually have `git`, `kitty`, and `nvim` configuration files in `~/.config`.

mislav replied:
`os.Remove`, like the `rm` command, is only able to remove empty directories. It would error out otherwise.

mislav commented on internal/config/config_file_test.go:
```diff
@@ -150,20 +153,149 @@ func Test_parseConfigFile(t *testing.T) {
+			env: map[string]string{
+				"GH_CONFIG_DIR": "/tmp/gh_config_dir",
+			},
+			output: ".tmp.gh_config_dir",
```
It's not really clear that this is a regexp pattern and that `.` is meant to match either style of slashes. You can use  static strings and`filepath.FromSlash()` here too to normalize path comparison across platforms

mislav commented on internal/config/config_file_test.go:
```diff
@@ -150,20 +153,149 @@ func Test_parseConfigFile(t *testing.T) {
+			name:   "no envVars",
+			output: ".config.gh",
+			env: map[string]string{
+				"AppData": "",
```
Shouldn't this test _always_ stub out all the environment variables that the implementation relies on? Meaning:
1. HOME
2. USERPROFILE
3. XDG_CONFIG_HOME
4. GH_CONFIG_DIR

This would be to neutralize the environment that the tests are run in. #3044

mislav reviewed (approved):
I'm fired up about this!

jasonkarns commented:
🎉 thanks @samcoe !!
